enhancements/is-bundle-proxying.md

```diff
@@ -0,0 +1,81 @@
+---
+title: ImageStream Bundle Proxying
+authors:
+  - "@njhale"
+reviewers:
+  - "@shawn-hurley"
+  - "@kevinrizza"
+approvers:
+  - "@benluddy"
+creation-date: 2020-06-25
+last-updated: 2020-06-25 
+status: provisional
+---
+
+# imagestream-bundle-proxying
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions
+
+- External images can be pulled into OpenShift’s integrated image registry
+- Pulling an external image to the integrated registry can be triggered by creating/updating an ImageStream resource
+- No non-standard cluster configuration is needed to allow external images to be pulled into the integrated registry
+- Pulls of external images into the registry abide by ImageContentSourcePolicy and other applicable cluster configurations
+- Direct -- non-pod spec -- pulling of images from the integrated registry transitively satisfies ImageContentSourcePolicy constraints
+
+## Summary
+
+When deployed to OpenShift, OLM can use an ImageStream to stage a bundle image in the internal image registry and access its image metadata while still meeting the OpenShift-specific constraint to pull images via CRI-O.
+
+## Motivation
+
+Today, when a bundle image is pulled and unpacked by OLM, a Job is created which specifies three containers:
+
+- a bundle util init container
+  - contains a statically linked bundle tool that can search the container’s filesystem for bundle metadata -- annotations.yaml -- and copy the referenced manifest files to a bundle volume
+  - mounts a single bundle util volume
+  - at runtime, uses cp to copy this bundle tool to the bundle util volume
+- a bundle data init container
+  - pull spec is used to delegate pulling of the bundle image via CRI-O
+  - mounts two volumes: bundle util and bundle data
+  - at runtime, executes the bundle tool from the bundle util mount to copy the bundle to the bundle data mount
+- a bundle extract container
+  - contains opm
+  - mounts the bundle data volume
+  - runs `opm extract` against the bundle data mount which results in a ConfigMap being created that contains the bundle’s kubectl-able content
+
+While this works for both OpenShift and upstream Kubernetes, the inability to inspect the image annotations -- where the path of the metadata directory is specified -- means that we cannot verify that the bundle metadata directory found in the bundle’s filesystem contains the data intended by the author.
+
+Using opm to directly interact with the image, instead of relying on a pod pull spec, allows for unambiguous resolution of the bundle metadata directory. In the future, this will also enable bundle data to be stored entirely as image annotations and/or manifest lists (w/o a run configuration or union filesystem).
```

Comment:
> allows for unambiguous resolution of the bundle metadata directory.

wouldn't just saying "this is always the metadata dir" also accomplish this?
or setting env vars on the image (which you can ref inside your container)?
or a metadata file inside the image in a well known location?

> In the future, this will also enable bundle data to be stored entirely as image annotations and/or manifest lists

why is storing bundle data entirely as image annotations a goal?  what's the use case?

Reply:
>wouldn't just saying "this is always the metadata dir" also accomplish this

This is actually something recently discussed. We may want to hard-code `metadata/annotations.yaml` as the metadata path.

>why is storing bundle data entirely as image annotations a goal

You would only ever have one source of truth for bundle metadata, which reduces ambiguity and the number of files required on-disk.

Reply:
I think this is fine for an iteration of the `registry+vX` spec for bundles, since we're tied to expressing bundles as standard container images. 

When we get custom mediatype support, I think we'll want to store manifests directly in manifest lists. At that point, we will no longer need to mount a filesystem, which comes with its own benefits (e.g. no longer need a node of the same platform to unpack a bundle built for it). I'm fairly sure this would still necessitate pulling bundles w/ `opm` directly instead of relying on a PodSpec.

Reply:
I think that preparing for a future that requires many prerequisites to be completed before we are able to use it while adding complexity for the clients and users of the system may not be the right trade-off in my opinion. I would think that if we get to the place where we can use the manifest list in that way, introducing a new type of bundle image seems reasonable.